Fleece/Support/ParseDate.cc

```diff
@@ -329,124 +309,294 @@ static void inject_local_tz(DateTime* p)
+        DateTime x{0, 0, 0, 0, 0, 0, 0, 0.0, 0, 0, 0, 0, 0};
+        if ( parseYyyyMmDd(dateStr, &x, false) ) { parseHhMmSs(dateStr, &x); }
+
+        return x;
+    }
+
+    DateTime ParseISO8601DateRaw(slice date) {
+        DateTime x{0, 0, 0, 0, 0, 0, 0, 0.0, 0, 0, 0, 0, 0};
+        auto     cstr = static_cast<char*>(malloc(date.size + 1));
+        if ( !cstr ) return x;
+
+        memcpy(cstr, date.buf, date.size);
+        cstr[date.size] = 0;
+        auto retVal     = ParseISO8601DateRaw(cstr);
+        free(cstr);
+        return retVal;
+    }
+
+    int64_t ToMillis(DateTime& dt, bool no_tz) {
+        if ( !dt.validHMS ) {
+            dt.h = dt.m = 0;
+            dt.s        = 0.0;
+            dt.validHMS = 1;
+        }
+
+        if ( !no_tz && !dt.validTZ ) { inject_local_tz(&dt); }
+
+        computeJD(&dt);
+        return dt.iJD - 210866760000000;
+    }
 
-    int64_t ParseISO8601Date(const char* zDate) {
+    DateTime FromMillis(int64_t timestamp) {
+        // Split out the milliseconds from the timestamp:
+        time_t secs{timestamp / 1000};
+        int    millis = (int)timestamp % 1000;
+
+        struct tm  timebuf {};
+        struct tm* result = gmtime_r(&secs, &timebuf);
+        return {0,
+                timebuf.tm_year + 1900,
+                timebuf.tm_mon + 1,
+                timebuf.tm_mday,
+                timebuf.tm_hour,
+                timebuf.tm_min,
+                0,
+                (double)timebuf.tm_sec + millis / 1000.0,
+                1,
+                1,
+                0,
+                1};
+    }
+
+    int64_t ParseISO8601Date(const char* dateStr) {
         DateTime x;
-        if (parseYyyyMmDd(zDate,&x))
-            return kInvalidDate;
-        computeJD(&x);
-        return x.iJD - 210866760000000;
+        if ( parseYyyyMmDd(dateStr, &x, true) ) return kInvalidDate;
+        return ToMillis(x);
     }
 
-    int64_t ParseISO8601Date(fleece::slice date) {
-        return ParseISO8601Date(string(date).c_str());
+    int64_t ParseISO8601Date(fleece::slice date) { return ParseISO8601Date(string(date).c_str()); }
+
+    DateComponent ParseDateComponent(slice component) {
+        string componentStr(component);
+
+        // Warning, tolower on negative char is UB so first convert to unsigned.  It doesn't matter
+        // if the result is nonsense since we are just using it as a lookup key
+        transform(componentStr.begin(), componentStr.end(), componentStr.begin(),
+                  [](unsigned char c) { return std::tolower(c); });
+        const auto entry = dateComponentMap.find(componentStr);
+        if ( entry == dateComponentMap.end() ) { return kDateComponentInvalid; }
+
+        return entry->second;
     }
 
-    slice FormatISO8601Date(char buf[], int64_t time, bool asUTC) {
-        if (time == kInvalidDate) {
+    slice FormatISO8601Date(char buf[], int64_t timestamp, bool asUTC, const DateTime* format) {
```

Comment:
Why do we need format? Does it mean that the caller may want to discard the YMD or HMS ?

Reply:
Yeah, per the PRD, we must support format other than ISO8601. We provide this in N1QL by allowing a DateTime parameter where the caller can pass a DateTime that looks like the format they want. i.e. if they pass "11:11:11" we would return them a datetime that only has HMS